main.tf

```diff
@@ -1,4 +1,7 @@
 locals {
+  # Store the list of principals and convert to set
+  ram_principals = var.create_tgw && var.share_tgw ? toset(var.ram_principals) : []
```

Comment:
simply moving this logic to a local does not correct its flaws

again, lets look at the conditional where its generally of the format `conditional ? true case : false case`
- When the conditional is true, the type is of `toset()`
- When the conditional is false, the type is of list since its just empty brackets

This is an issue because the types are now mis-matched (`toset()` vs list). 

In addition, because you are using `toset()` on a list, any of the values in that list cannot be computed or else you will face the well known issue of not being able to use `for_each` on a value that is not yet known. 

Lastly, changing from count to for_each is a breaking change

Reply:
I guess what @bryantbiggs is trying to say is since the left side of the conditional is a `set` you need the right one to be set too, i.e. `toset([])` instead of `[]` which is of type `list`.

Reply:
> I guess what @bryantbiggs is trying to say is since the left side of the conditional is a `set` you need the right one to be set too, i.e. `toset([])` instead of `[]` which is of type `list`.

Exactly, Igor! Thanks 🙇 

However, even solving this part (making the sides equal) we will face that `for_each` known issue mentioned by Bryan... 😢

Reply:
I was take a looking on `main.tf` again and currently it uses `count` on `length(var.ram_principals)` here:
https://github.com/terraform-aws-modules/terraform-aws-transit-gateway/blob/f4db6678b5c67fbecb5d50397721d8d363bc3de7/main.tf#L165

How `count` will know the length of the list for computed values? In this case we wil face the same problem (unknown value), right?